src/main/scala/lc/core/captcha.scala

```diff
@@ -113,23 +96,30 @@ class Captcha {
   }
 
   def checkAnswer(answer: Answer): Result = {
-    val selectPstmt = Statements.tlStmts.get.selectPstmt
-    selectPstmt.setInt(1, Config.captchaExpiryTimeLimit)
-    selectPstmt.setString(2, answer.id)
-    val rs: ResultSet = selectPstmt.executeQuery()
-    val psOpt = if (rs.first()) {
-      val secret = rs.getString("secret")
-      val provider = rs.getString("provider")
-      val check = CaptchaProviders.getProviderById(provider).checkAnswer(secret, answer.answer)
-      val result = if (check) ResultEnum.TRUE.toString else ResultEnum.FALSE.toString
-      result
-    } else {
-      ResultEnum.EXPIRED.toString
+    try{
+      val selectPstmt = Statements.tlStmts.get.selectPstmt
+      selectPstmt.setInt(1, Config.captchaExpiryTimeLimit)
+      selectPstmt.setString(2, answer.id)
+      val rs: ResultSet = selectPstmt.executeQuery()
+      val psOpt = if (rs.first()) {
+        val secret = rs.getString("secret")
+        val provider = rs.getString("provider")
+        val check = CaptchaProviders.getProviderById(provider).checkAnswer(secret, answer.answer)
+        val result = if (check) ResultEnum.TRUE.toString else ResultEnum.FALSE.toString
+        result
+      } else {
+        ResultEnum.EXPIRED.toString
+      }
+      val deleteAnswerPstmt = Statements.tlStmts.get.deleteAnswerPstmt
+      deleteAnswerPstmt.setString(1, answer.id)
+      deleteAnswerPstmt.executeUpdate()
+      Result(psOpt)
+    } catch {
+      case exception: Exception => {
+        println(exception.getStackTrace)
+        throw new Exception(exception.getMessage)
```

Comment:
You are rethrowing the exception. Is this only for debugging?

Actually, I liked  the earlier design better. Having an error type, ensure that compiler can check that we are handling the error path. Exception in scala are not checked, unlike in Java.

My suggestion is to catch the exception at the highest level:

```
def makeApiWorker(path: String, f: (method: String, ex: Exchange) => Response) : Steretypes.HttpWorker {
    new Stereotypes.HttpWorker(path, ex => {
      val requestMethod = ex.getRequestMethod
      val response = try {
          f(requestMethod, ex)
      } catch {
        case exception: Exception =>{
            getException(exception)
          }
        }
      }
      sendResponse(statusCode = response.statusCode, message = response.message, ex = ex, media = false)
}
```

And for each API endpoint we can do something like this:

```scala
makeApiWorker("/v1/captcha", (method, ex) => {
      if (method == "POST"){
          val json = getRequestJson(ex)
          val param = json.extract[Parameters]
          val id = captcha.getChallenge(param)
          Response(200, write(id))
      } else {
        getBadRequestError()
      }
    }),
```

So the exception handling and response sending need not be repeated everywhere.

Reply:
The shortcoming of the earlier design was that we were returning an `Error` case class, that wasn't really an error and hence would not be caught at any level. Moreover, since it wasn't an error, the API returned the correct error message while the status code was still 200.

Reply:
> we were returning an Error case class, that wasn't really an error and hence would not be caught at any level

I didn't understand this. Let's discuss this on call.